### PR TITLE
Make analyzer units into table test

### PR DESCRIPTION
**Pull request by natalieparellano**
Also remove platform from layer metadata restorer because it's not used.

Signed-off-by: Natalie Arellano <narellano@vmware.com>

### CODE REVIEW

File: analyzer_test.go [hunk 1]
```diff
@@ -1,13 +1,650 @@
...
 
 func TestAnalyzer(t *testing.T) {
-	spec.Run(t, "Analyzer", testAnalyzer06, spec.Report(report.Terminal{}))
-	spec.Run(t, "Analyzer", testAnalyzer07, spec.Report(report.Terminal{}))
+	for _, api := range api.Platform.Supported {
+		spec.Run(t, "unit-analyzer/"+api.String(), testAnalyzerBuilder(api.String()), spec.Parallel(), spec.Report(report.Terminal{}))
+	}
+}
+
+func testAnalyzerBuilder(apiString string) func(t *testing.T, when spec.G, it spec.S) {
```
Comment (natalieparellano) [inline analyzer_test.go:35]: To get a better idea of the changes, it might be best to compare this with main.

**Comment by natalieparellano** [comment]
Pending un-draft on answering the TODO

File: layer_metadata_restorer_test.go [hunk 2]
```diff
@@ -40,8 +40,7 @@ func testLayerMetadataRestorer(t *testing.T, when spec.G, it spec.S) {
 		layerDir, err = ioutil.TempDir("", "lifecycle-layer-dir")
 		h.AssertNil(t, err)
 		logger := &log.Logger{Handler: &discard.Handler{}}
-		platform := platform.NewPlatform("0.7")
-		layerAnalyzer = lifecycle.NewLayerMetadataRestorer(logger, layerDir, platform, skipLayers)
+		layerAnalyzer = lifecycle.NewLayerMetadataRestorer(logger, layerDir, skipLayers)
```
Comment (yaelharel) [inline layer_metadata_restorer_test.go:43]: Just to make sure I'm not missing something: we're planning to delete `restorer_platform06_test.go` and `restorer_platform07_test.go`, right?

Comment (natalieparellano) [reply layer_metadata_restorer_test.go:43]: Yeah that'd be next.

Comment (natalieparellano) [reply layer_metadata_restorer_test.go:43]: There's also this issue which is related: https://github.com/buildpacks/lifecycle/issues/587

File: analyzer_test.go [hunk 3]
```diff
@@ -1,13 +1,649 @@
...
+	return func(t *testing.T, when spec.G, it spec.S) {
+		var (
+			analyzer         *lifecycle.Analyzer
+			cacheDir         string
```
Comment (yaelharel) [inline analyzer_test.go:39]: It's not really important but I know that usually we're trying to put all of the variables of the same type one after the other.

File: analyzer_test.go [hunk 4]
```diff
@@ -1,13 +1,649 @@
...
+	"github.com/buildpacks/lifecycle/cmd"
+	"github.com/buildpacks/lifecycle/platform"
+	h "github.com/buildpacks/lifecycle/testhelpers"
+	"github.com/buildpacks/lifecycle/testmock"
 )
 
 func TestAnalyzer(t *testing.T) {
-	spec.Run(t, "Analyzer", testAnalyzer06, spec.Report(report.Terminal{}))
-	spec.Run(t, "Analyzer", testAnalyzer07, spec.Report(report.Terminal{}))
+	for _, api := range api.Platform.Supported {
```
Comment (yaelharel) [inline analyzer_test.go:30]:
```suggestion
	for _, platformApi := range api.Platform.Supported {
```

File: analyzer_test.go [hunk 5]
```diff
@@ -1,13 +1,649 @@
...
+		spec.Run(t, "unit-analyzer/"+api.String(), testAnalyzerBuilder(api.String()), spec.Parallel(), spec.Report(report.Terminal{}))
+	}
+}
+
+func testAnalyzerBuilder(apiString string) func(t *testing.T, when spec.G, it spec.S) {
```
Comment (yaelharel) [inline analyzer_test.go:35]:
```suggestion
func testAnalyzerBuilder(platformApiString string) func(t *testing.T, when spec.G, it spec.S) {
```

File: analyzer_test.go [hunk 6]
```diff
@@ -1,13 +1,649 @@
...
+				when("buildpack API < 0.6", func() {
+					it.Before(func() {
+						analyzer.Buildpacks = []buildpack.GroupBuildpack{
+							{ID: "metadata.buildpack", API: "0.5"},
+							{ID: "no.cache.buildpack", API: "0.5"},
+						}
+					})
+
+					it("restores layer metadata and doesn't change the values of the launch, build and cache flags in top level", func() {
+						h.SkipIf(t, api.MustParse(apiString).Compare(api.MustParse("0.7")) >= 0, "Platform API >= 0.7 does not restore layer metadata")
```
Comment (yaelharel) [inline analyzer_test.go:149]: Can we move this line to the `it.Before` on line 141?

File: analyzer_test.go [hunk 7]
```diff
@@ -1,13 +1,649 @@
...
+
+						path := filepath.Join(layersDir, "no.cache.buildpack", "some-layer.toml")
+						got := h.MustReadFile(t, path)
+						want := "[metadata]\n  some-layer-key = \"some-layer-value\""
+
+						h.AssertStringContains(t, string(got), want)
+					})
+
+					it("does not restore layers for undetected buildpacks", func() {
+						h.SkipIf(t, api.MustParse(apiString).Compare(api.MustParse("0.7")) >= 0, "Platform API >= 0.7 does not restore layer metadata")
```
Comment (yaelharel) [inline analyzer_test.go:221]: Please see https://github.com/buildpacks/lifecycle/pull/588/files#r618456024

File: analyzer_test.go (hunk 6 above)
Comment (jabrown85) [reply analyzer_test.go:149]: IIRC `SkipIf` has to be inside an `it` block.

File: analyzer_test.go [hunk 8]
```diff
@@ -1,13 +1,649 @@
...
+						{"metadata.buildpack/store.toml", "[metadata]\n  [metadata.metadata-buildpack-store-data]\n    store-key = \"store-val\""},
+						{"no.cache.buildpack/store.toml", "[metadata]\n  [metadata.no-cache-buildpack-store-data]\n    store-key = \"store-val\""},
+					} {
+						got := h.MustReadFile(t, filepath.Join(layersDir, data.name))
+						h.AssertStringContains(t, string(got), data.want)
+					}
+				})
+
+				when("cache exists", func() {
+					it.Before(func() {
```
Comment (yaelharel) [inline analyzer_test.go:254]: I think you can put the `h.SkipIf` here instead of all of the nested `it` and `when`.

**Review by yaelharel — APPROVED** [review]
The PR looks good!
I put some comments on it.

Regarding my comments about moving the `h.SkipIf` to the `it.Before`:
I'm not sure if we should move it. My initial thought was to reduce the number of `SkipIf` calls and to not run the `it.Before` steps if they're not needed. But it might be clearer to have it written down next to each `it` for better readability and in case we'll add more tests under `when`.

File: analyzer_test.go (hunk 6 above)
Comment (yaelharel) [reply analyzer_test.go:149]: Interesting. I was able to put it under `it.Before`.

**Pull request merged**